File: dotnet/autonomous/github-trending/Tools/GitHubTrendingTool.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System.ComponentModel;
using System.Text;
using System.Text.Json;
using Microsoft.Agents.A365.Observability.Runtime.Tracing.Contracts; // A365 Observability — ToolCallDetails, AgentDetails, etc.
using Microsoft.Agents.A365.Observability.Runtime.Tracing.Scopes;    // A365 Observability — ExecuteToolScope

namespace GitHubTrending.Tools;

/// <summary>
/// Tool that fetches trending repositories from the GitHub Search API.
/// Uses the unauthenticated search endpoint (no API key required, 10 req/min rate limit).
/// </summary>
public sealed class GitHubTrendingTool
{
    private static readonly Uri GitHubApiEndpoint = new("https://api.github.com");

    private readonly HttpClient _httpClient;
    private readonly string _language;
    private readonly int _minStars;
    private readonly int _maxResults;

    // A365 Observability — injected from Agent365ObservabilityContext
    private readonly AgentDetails _agentDetails;

    public GitHubTrendingTool(
        IHttpClientFactory httpClientFactory,
        IConfiguration configuration,
        Agent365ObservabilityContext obsContext)
    {
        _httpClient = httpClientFactory.CreateClient();
        _httpClient.DefaultRequestHeaders.UserAgent.ParseAdd("GitHubTrendingAgent/1.0");

        _language = configuration["GitHubTrending:Language"] ?? "csharp";
        _minStars = configuration.GetValue<int>("GitHubTrending:MinStars", 5);
        _maxResults = configuration.GetValue<int>("GitHubTrending:MaxResults", 10);

        _agentDetails = obsContext.AgentDetails;
    }

    [Description("Search GitHub for repositories created in the last 7 days that are trending by star count")]
    public async Task<string> GetTrendingRepositories(
        [Description("Optional programming language filter (e.g. 'csharp', 'python', 'typescript'). Leave empty for all languages.")]
        string? language = null)
    {
        var lang = language ?? _language;

        // A365 Observability — ExecuteTool span wraps the GitHub API call
        var request = new Request(content: language);
        using var toolScope = ExecuteToolScope.Start(
            request,
            new ToolCallDetails(
                toolName: nameof(GetTrendingRepositories),
                arguments: language,
                toolType: ToolType.Function,
                description: "Search GitHub for trending repositories by star count",
                endpoint: GitHubApiEndpoint),
            _agentDetails);

        var since = DateTime.UtcNow.AddDays(-7).ToString("yyyy-MM-dd");

        var query = $"created:>{since} stars:>={_minStars}";
        if (!string.IsNullOrWhiteSpace(lang))
            query += $" language:{lang}";

        var url = $"https://api.github.com/search/repositories" +
                  $"?q={Uri.EscapeDataString(query)}" +
                  $"&sort=stars&order=desc&per_page={_maxResults}";

        using var response = await _httpClient.GetAsync(url).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            var errorResult = $"GitHub API request failed: HTTP {response.StatusCode}";
            toolScope.RecordResponse(errorResult);
            return errorResult;
        }

        using var doc = await JsonDocument.ParseAsync(
            await response.Content.ReadAsStreamAsync().ConfigureAwait(false)).ConfigureAwait(false);

        var items = doc.RootElement.GetProperty("items");
        if (items.GetArrayLength() == 0)
        {
            var emptyResult = $"No trending repositories found for language '{lang}' in the last 7 days.";
            toolScope.RecordResponse(emptyResult);
            return emptyResult;
        }

        var sb = new StringBuilder();
        sb.AppendLine($"Top {Math.Min(items.GetArrayLength(), _maxResults)} trending {lang} repositories (created after {since}):");
        sb.AppendLine();

        foreach (var repo in items.EnumerateArray())
        {
            var name = repo.GetProperty("full_name").GetString();
            var stars = repo.GetProperty("stargazers_count").GetInt32();
            var description = repo.TryGetProperty("description", out var desc) && desc.ValueKind != JsonValueKind.Null
                ? desc.GetString() ?? "(no description)"
                : "(no description)";
            var htmlUrl = repo.GetProperty("html_url").GetString();

            sb.AppendLine($"- **{name}** ({stars} stars): {description}");
            sb.AppendLine($"  {htmlUrl}");
        }

        var result = sb.ToString();
        toolScope.RecordResponse(result);
        return result;
    }
}

File: nodejs/autonomous/github-trending/src/tools/github-trending-tool.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

/**
 * Tool that fetches trending repositories from the GitHub Search API.
 * Uses the unauthenticated search endpoint (no API key required, 10 req/min rate limit).
 */

import {
  AgentDetails,
  ExecuteToolScope,
  Request,
  ToolCallDetails,
  ServiceEndpoint,
} from '@microsoft/agents-a365-observability';

const GITHUB_API_ENDPOINT: ServiceEndpoint = { host: 'api.github.com', protocol: 'https' };

export async function getTrendingRepositories(
  agentDetails: AgentDetails,
  language: string = 'typescript',
  minStars: number = 5,
  maxResults: number = 10,
): Promise<string> {
  // A365 Observability — ExecuteTool span wraps the GitHub API call
  const request: Request = { content: language };
  const toolDetails: ToolCallDetails = {
    toolName: 'get_trending_repositories',
    arguments: language,
    toolType: 'function',
    description: 'Search GitHub for trending repositories by star count',
    endpoint: GITHUB_API_ENDPOINT,
  };

  const scope = ExecuteToolScope.start(request, toolDetails, agentDetails);
  try {
    return await scope.withActiveSpanAsync(async () => {
      const since = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString().split('T')[0];

      let query = `created:>${since} stars:>=${minStars}`;
      if (language) {
        query += ` language:${language}`;
      }

      const url = `https://api.github.com/search/repositories?q=${encodeURIComponent(query)}&sort=stars&order=desc&per_page=${maxResults}`;

      const response = await fetch(url, {
        headers: { 'User-Agent': 'GitHubTrendingAgent/1.0' },
      });

      if (!response.ok) {
        const errorResult = `GitHub API request failed: HTTP ${response.status}`;
        scope.recordResponse(errorResult);
        return errorResult;
      }

      const data = await response.json() as { items?: Array<{ full_name: string; stargazers_count: number; description?: string; html_url: string }> };
      const items = data.items || [];

      if (items.length === 0) {
        const emptyResult = `No trending repositories found for language '${language}' in the last 7 days.`;
        scope.recordResponse(emptyResult);
        return emptyResult;
      }

      const lines: string[] = [
        `Top ${Math.min(items.length, maxResults)} trending ${language} repositories (created after ${since}):`,
        '',
      ];

      for (const repo of items.slice(0, maxResults)) {
        const description = repo.description || '(no description)';
        lines.push(`- **${repo.full_name}** (${repo.stargazers_count} stars): ${description}`);
        lines.push(`  ${repo.html_url}`);
      }

      const result = lines.join('\n');
      scope.recordResponse(result);
      return result;
    });
  } finally {
    scope.dispose();
  }
}

/**
 * Tool definition for OpenAI function calling.
 */
export const TOOL_DEFINITION = {
  type: 'function' as const,
  function: {
    name: 'get_trending_repositories',
    description: 'Search GitHub for repositories created in the last 7 days that are trending by star count',
    parameters: {
      type: 'object',
      properties: {
        language: {
          type: 'string',
          description: "Optional programming language filter (e.g. 'typescript', 'python', 'csharp'). Leave empty for all languages.",
        },
      },
    },
  },
};
